### PR TITLE
Feature/sc 412

### PR DESCRIPTION
**Pull request by CaptainOfHacks**

### CODE REVIEW

File: sem_covid/adapters/embedding_models.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                for word in tokens]
+
+
+class AverageSentenceEmbeddingModel(SentenceEmbeddingModelABC):
+
+    def __init__(self, word_embedding_model: WordEmbeddingModelABC, tokenizer: TokenizerModelABC):
+        self._word_embedding_model = word_embedding_model
+        self._tokenizer = tokenizer
+
+    def encode(self, sentences: List[str]) -> List:
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:53]: Let's also add nice and explicit documentation of what this class is intended for.

I believe here we want that the list of `tokens` is in fact a set of consecutive tokens of a sentence or a paragraph or a document. 

Possibly as are in the text or possibly tokens after the stop-words and punctuation were removed. Having this list instead of the original text is an advantage because we have externalised tokenisation and pre-processing.

File: sem_covid/adapters/embedding_models.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        results = []
+        for sentence in sentences:
+            embeddings = self._word_embedding_model.encode(self._tokenizer.tokenize(sentence))
+            results += [np.mean([np.zeros(len(embeddings[0])).tolist()] + embeddings, axis=0).tolist()]
+        return results
+
+
+class TfIdfSentenceEmbeddingModel(AverageSentenceEmbeddingModel):
+
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:62]: lets document this class, what are the intentions and how it should function;
and how it should be used :)

File: sem_covid/adapters/embedding_models.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        results = []
+        for index in range(0, len(tokens)):
+            if tokens[index] in tf_idf_row.index:
+                weight = tf_idf_row[tokens[index]]
+                sum_weights += weight
+                results += [(weight * np.array(embeddings[index])).tolist()]
+
+        return (np.sum(results, axis=0) / sum_weights).tolist()
+
+    def encode(self, sentences: List[str]) -> List:
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:76]: document what it does

File: sem_covid/adapters/embedding_models.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        tf_idf_matrix = pd.DataFrame(tf_idf_vectors.fit_transform(sentences).todense().tolist(),
+                                     columns=tf_idf_vectors.get_feature_names()
+                                     )
+        return [
+            self.encode_one_sentence(sentences[index], tf_idf_matrix.iloc[index])
+            for index in range(0, len(sentences))
+        ]
+
+
+class UniversalSentenceEmbeddingModel(SentenceEmbeddingModelABC):
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:87]: document the class, and the intention behind it, and how it works

File: sem_covid/adapters/embedding_models.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self):
+        model_url = "https://tfhub.dev/google/universal-sentence-encoder/4"
+        self.model = hub.load(model_url)
+
+    def encode(self, sentences: List[str]) -> List:
+        return self.model(sentences).numpy().tolist()
+
+
+class EurLexBertSentenceEmbeddingModel(SentenceEmbeddingModelABC):
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:97]: document, intention (input-ouput), and internal mechanisms

File: sem_covid/adapters/embedding_models.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self):
+        self.tokenizer = AutoTokenizer.from_pretrained("nlpaueb/bert-base-uncased-eurlex")
+        self.model = AutoModel.from_pretrained("nlpaueb/bert-base-uncased-eurlex")
+
+    def encode(self, sentences: List[str]) -> List:
```
Comment (costezki) [inline sem_covid/adapters/embedding_models.py:103]: document how should it function?

File: tests/unit/test_embedding_models/__init__.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
```
Comment (costezki) [inline tests/unit/test_embedding_models/__init__.py:1]: Let's stick to this headder 

https://meaningfy.atlassian.net/wiki/spaces/EP/pages/370671617/Python+header+template

File: tests/unit/test_embedding_models/test_sentence_embedding_model.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+from sem_covid.adapters.abstract_model import SentenceEmbeddingModelABC, WordEmbeddingModelABC
+from sem_covid.adapters.embedding_models import (AverageSentenceEmbeddingModel, Word2VecEmbeddingModel,
+                                                 BasicTokenizerModel, SpacyTokenizerModel, TfIdfSentenceEmbeddingModel,
+                                                 UniversalSentenceEmbeddingModel,
+                                                 EurLexBertSentenceEmbeddingModel)
+from tests.unit.conftest import nlp
+from tests.unit.test_embedding_models.fake_keyed_vectors import FakeKeyedVectors
+
+SENTENCES = [
+    'Rather than having all of its functionality built into its core.',
```
Comment (costezki) [inline tests/unit/test_embedding_models/test_sentence_embedding_model.py:15]: Nice examples :)

File: tests/unit/test_embedding_models/test_sentence_embedding_model.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    )
+    check_sentence_embedding_model(tf_idf_sentence_embedding_model_basic)
+    tf_idf_sentence_embedding_model_spacy = TfIdfSentenceEmbeddingModel(
+        word_embedding_model=get_word2vec_emb_model(),
+        tokenizer=SpacyTokenizerModel(spacy_nlp=nlp)
+    )
+    check_sentence_embedding_model(tf_idf_sentence_embedding_model_spacy)
+
+
+def test_universal_sentence_embedding_model():
```
Comment (costezki) [inline tests/unit/test_embedding_models/test_sentence_embedding_model.py:68]: Good!

File: tests/unit/test_embedding_models/test_sentence_embedding_model.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+SENTENCES = [
+    'Rather than having all of its functionality built into its core.',
+    'Python was designed to be highly extensible (with modules).',
+    'Rather than having all of its functionality built into its core.',
+    'This compact modularity has made it particularly popular as a means of adding programmable interfaces to existing applications.',
+    "Van Rossum's vision of a small core language with a large standard library and easily extensible interpreter stemmed from his frustrations with ABC, which espoused the opposite approach."
+]
+
+
+def check_sentence_embedding_model(sentence_embedding_model: SentenceEmbeddingModelABC):
```
Comment (costezki) [inline tests/unit/test_embedding_models/test_sentence_embedding_model.py:23]: I like helper test functions.
They are best placed in `conftest.py` to maximise reuse. 

Unless they are very very local :)

File: tests/unit/test_embedding_models/test_tokenizer_model.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    tokens = tokenizer.tokenize(text)
+    assert len(tokens) == 5
+    assert tokens[0] == 'sample'
+    assert tokens[4] == 'test'
+
+
+def test_spacy_tokenizer_model():
+    tokenizer = SpacyTokenizerModel(spacy_nlp=nlp)
+    assert isinstance(tokenizer, TokenizerModelABC)
+    text = 'sample text for this test'
```
Comment (costezki) [inline tests/unit/test_embedding_models/test_tokenizer_model.py:24]: Here you can use text like this

```
len(tokenise("Don't throw morphemes away" )) == 5
```

**Review by costezki — APPROVED** [review]
Very nice set of tools. 
We can discuss a bit their naming conventions and final intention behind each of these tools, but a very nice effort overall. We need these tools.

File: sem_covid/adapters/embedding_models.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self, word_embedding_model: WordEmbeddingModelABC, tokenizer: TokenizerModelABC):
+        self._word_embedding_model = word_embedding_model
+        self._tokenizer = tokenizer
+
+    def encode(self, sentences: List[str]) -> List:
+        results = []
+        for sentence in sentences:
+            embeddings = self._word_embedding_model.encode(self._tokenizer.tokenize(sentence))
+            results += [np.mean([np.zeros(len(embeddings[0])).tolist()] + embeddings, axis=0).tolist()]
```
[the remark this answers is not in the document]
Comment (CaptainOfHacks) [reply sem_covid/adapters/embedding_models.py:59]: Q :"Why tokenise and include the average function here? Don't we have it already above? Possibly this class is not very useful."
A: this was done in order to standardize all approaches, the old implementation will be replaced over time with this new one.
